File: typescript/bob/bob.ts
class Bob {
  message = "";

  hey(message: string): string {
    this.message = message.trim();

    if (this.empty()) {
      return "Fine. Be that way!";
    }

    if (this.allUpperCase() && this.askingQuestion()) {
      return "Calm down, I know what I'm doing!";
    }

    if (this.askingQuestion()) {
      return "Sure.";
    }

    if (this.allUpperCase() && !this.noLetters()) {
      return "Whoa, chill out!";
    }

    return "Whatever.";
  }

  private empty(): boolean {
    return this.message == "";
  }

  private allUpperCase(): boolean {
    return this.message == this.message.toUpperCase() && !this.noLetters();
  }

  private askingQuestion(): boolean {
    return this.message[this.message.length - 1] == "?"
  }

  private noLetters(): boolean {
    return !(/[a-zA-Z]/.test(this.message));
  }
}

export default Bob

File: ruby/bob/bob.rb
class Bob

  def self.hey(message)
    message = message.strip

    if being_shouted_at?(message) && being_questioned?(message)
      "Calm down, I know what I'm doing!"
    elsif being_shouted_at?(message)
      "Whoa, chill out!"
    elsif being_questioned?(message)
      "Sure."
    elsif given_silent_treatment?(message)
      "Fine. Be that way!"
    else
      "Whatever."
    end
  end

  private

  def self.being_questioned?(message)
    message.end_with?("?")
  end

  def self.being_shouted_at?(message)
    contains_letters?(message) && message == message.upcase
  end

  def self.given_silent_treatment?(message)
    message.strip == ""
  end

  def self.contains_letters?(message)
    message.scan(/[A-Za-z]/).any?
  end

end
